doc/source/DBEP/DBEP02.rst

```diff
@@ -0,0 +1,167 @@
+need to coordinate in any way (or even be aware of each others
+existence). In contrast, the *Resource* and *Datum* documents are
+inserted directly into an *AssetRegistry* by the *ophyd* objects.
+This breaks the separation we have between the data collection process
+/ hardware, the generation of the documents, and the consumers of
+those documents and leads to several unfortunate situations:
+
+ - *ohpyd* objects hold an instance of an *AssetRegisty*
+ - we need to keep track of _which_ *AssertRegistry* things were
+   inserted into
+ - consumers that want access to the asset documents need to also have
+   a handle to the database that the objects are inserting into
+
+The proposed solution is to promote *Resource* and *Datum* documents
+to be peer documents with *Start*, *Stop*, *Descriptor* and *Event*.
+They will appear in the document stream and be inserted into
+*DataBroker* via ``db.insert``.  This eliminates the 'special'
+side-band communication and brings all consumers back to the same
+footing.  This will require coordinated changes to *event-model*,
+*databroker*, *bluesky*, and *ophyd*.
+
+
+
+Implementation
+==============
+
+Currently, *ophyd* is responsible for collecting all of the values for
+the *Resource* and *Datum* documents except for the uids.  The uids
+are generated by calls to ``reg.register_*`` and the datum uids are
+subsequently returned to the *RunEngine* via ``obj.read``.  The
+proposed change is:
+
+ 1. *ophyd* objects would be responsible for generating the full *Resource*
+    and *Datum* documents and providing them to the *RunEngine* to be
+    emitted (likely should provide some helpers).
+
+    a. Similar to the current documents, a *Resource* must be emitted
+       before any *Datum* that refers to it per *Start* / *Stop* pair
+    b. an identical (including uid) *Resource* maybe emitted more than
+       once, the consumers will need to handle this.
+    c. an identical *Datum* document may be returned the *RunEngine*
+       more than once.  Do we want to allow the RE to emit them more
+       than once? **[QUESTION]**
+
+ 2. in ``save`` before the *Event* document is emitted the *RunEngine*
+    will acquire and emit any *AssetRegistry* Documents.
+
+    a. in ``save`` the *RunEngine* knows what objects in the bundle, call
+       ``get_asset_document_gen`` **[RENAME]** method ::
+
+            def gadg(self) -> Iterator[Tuple[str, Dict[str, Any]]]:
+                ...
+
+       which will yield the ``(name, doc)`` pairs for anything that
+       was just read.
+
+    b. these documents will be emitted **before** the *Event*
+
+ 3. consumers will now have access to all relevant documents and can
+    do what ever they want with them (insert into an asset registry,
+    live processing / display, copy files else where)
+
+event-model
+-----------
+
+ 1. add schema for *Resource* and *Datum*
+ 2. assert that datum_id must be of the form ``{resource_id}/{N}``.
```

Comment:
I don't know if this was the previous way of writing it but this sounds very sensible.

Reply:
previously there was no constraints so it was just uuid4 values.